### ISSUE TITLE
RPC schemas work differently in local supabase dev environment, causing code to break only after live deployment

### ISSUE THREAD

**Issue by dawnmist** [supabase/supabase#20123]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I have been developing my application mostly within the schema name 'app', since I didn't want public graphql endpoints, etc.

When I implemented a few rpc calls, those calls to `app.max_participant_count()` work against the local supabase dev environment.

When I pushed the code using those rpc calls to my staging environment, those rpc calls now generate the error:
```json
"error":{
  "code":"PGRST106",
  "details":null,
  "hint":null,
  "message":"The schema must be one of the following: public, storage, graphql_public"
}
```

If using rpc is not permitted on custom schemas, the local dev environment _should_ reflect this restriction so the schema issue can be resolved at development time, rather than only discovering the restriction _after_ the code has been deployed to a live environment and breaks.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Create a migration that creates a function in a custom schema
```sql
create
or replace function app.current_participant_count () returns int as $$
  select count(p.id) from app.participants p
  where p.owner_id = auth.uid();
$$ language sql stable security definer;
```
2. Use `supabase db reset` to reset your local db to include the new migration
3. Use the rpc call in your code, note that it works perfectly well in local development
```ts
// supabase server client creation
import { createServerClient } from '@supabase/auth-helpers-remix';
import type { Database } from './schema.ts';

export const createServerSupabase = (request: Request, response: Response) =>
  createServerClient<Database, 'app'>(ENV.SUPABASE_URL, ENV.SUPABASE_ANON_KEY, {
    request,
    response,
    options: {
      db: {
        schema: 'app',
      },
    },
  });

// supabase access function
export const numOwnedParticipants = async (supabase: ReturnType<typeof createServerSupabase>) => {
  const { data, error } = await supabase.rpc('current_participant_count');

  if (error) {
    const { data: sessionData } = await supabase.auth.getSession();
    const session = sessionData?.session;
    logger.error({
      msg: 'Error getting participant count',
      error,
      userId: session?.user?.id,
      userEmail: session?.user?.email,
    });
    throw error;
  }

  return data;
};

// use within remix loader  
export const loader = async ({ request }: LoaderFunctionArgs) => {
  const response = new Response();
  const session = await requireSession({ request, response });

  const supabase = createServerSupabase(request, response);
  const numParticipants = await numOwnedParticipants(supabase);

  return json({ numParticipants }, { headers: response.headers });
};
```
4. Connect instead to a supabase hosted db instance & push the migration up so the new function is in the supabase db.
5. Note that the exact same code is now rejected by the supabase hosted db due to the function being defined in a custom schema - even though your supabase client was created specifically for that custom schema as per the snippet above and the function had been defined in the supabase client's datatypes.

## Expected behavior

If rpc calls to a custom schema are banned, the local dev environment should **also** ban them so that the restriction can be discovered **before** the code using those functions goes live.

## System information

- OS: Debian 12 x86_64
- Browser: Google Chrome 120.0.6099.129
- Version of supabase-js:
    "@supabase/auth-helpers-remix": "0.2.6",
    "@supabase/supabase-js": "2.38.5",
    "supabase": "1.115.4",
- Version of Node.js: v20.10.0

**Comment by GaryAustin1** [comment]
Did you add the schema to the API settings here?
![image](https://github.com/supabase/supabase/assets/54564956/2a37cc9e-465d-4a2f-b998-7682b15deeed)

**Comment by dawnmist** [comment]
You're correct - I hadn't remembered/found that setting when I ran into the problem. It doesn't exist in local dev, and I hadn't seen it when I was looking for why things had broken after deployment. The error message hadn't indicated that the schema list was configurable either, so I'd taken the error as a fixed/unchangeable one.

Follow-up question - the reason I'd originally used a different schema was because I didn't want a graphql schema published with the table datatypes publicly accessible. Does adding the schema containing the rpc functions to the "Exposed Schemas" list result in the tables in that schema being added to a public graphql schema as well?
